engines/sci/util.h

```diff
@@ -45,6 +50,715 @@ void WRITE_SCI11ENDIAN_UINT32(void *ptr, uint32 val);
+			operator+=(1);
+			return old;
+		}
+
+		inline SpanIterator &operator--() {
+			return operator+=(-1);
+		}
+
+		inline SpanIterator operator--(int) {
+			SpanIterator old = *this;
+			operator+=(-1);
+			return old;
+		}
+
+		inline const SpanIterator operator+(const index_type delta) const {
+			SpanIterator it = *this;
+			return it += delta;
+		}
+
+		inline SpanIterator operator+(const index_type delta) {
+			SpanIterator it = *this;
+			return it += delta;
+		}
+
+		inline const SpanIterator operator-(const index_type delta) const {
+			return operator+(-delta);
+		}
+
+		inline SpanIterator operator-(const index_type delta) {
+			return operator+(-delta);
+		}
+
+		inline index_type operator-(const SpanIterator &other) const {
+			assert(_span == other._span);
+			return _index - other._index;
+		}
+
+#pragma mark -
+#pragma mark SpanIterator - Comparison operations
+
+		inline bool operator==(const SpanIterator& other) const {
+			return _span == other._span && _index == other._index;
+		}
+
+		inline bool operator!=(const SpanIterator& other) const {
+			return !operator==(other);
+		}
+
+		inline bool operator<(const SpanIterator& other) const {
+			assert(_span == other._span);
+			return _index < other._index;
+		}
+
+		inline bool operator<=(const SpanIterator& other) const {
+			return !other.operator<(*this);
+		}
+
+		inline bool operator>(const SpanIterator& other) const {
+			return other.operator<(*this);
+		}
+
+		inline bool operator>=(const SpanIterator& other) const {
+			return !operator<(other);
+		}
+
+	protected:
+		container_type *_span;
+		index_type _index;
+	};
+} // end of namespace SpanInternal
+
+#pragma mark -
+#pragma mark SpanOwner
+
+/**
+ * Similar to DisposablePtr, but allows holding and disposing pointers created
+ * by malloc, and provides convenience functions for Span creation from files,
+ * streams, and other Spans.
+ */
+template <typename Span>
+class SpanOwner {
+public:
+	typedef typename Span::value_type value_type;
+	typedef uint32 index_type;
+	typedef uint32 size_type;
+
+	inline SpanOwner() : _span() {}
+
+	/**
+	 * Creates a new owned span of uninitialized memory of the given size.
+	 */
+	inline SpanOwner(const size_type dataSize, const Common::String &sourceName = Common::String()) {
+		allocate(dataSize, sourceName);
+	}
+
+	/**
+	 * Creates a new owned span from the data in the given file.
+	 */
+	inline SpanOwner(Common::File &file, const size_type dataSize = kSpanMaxSize) {
+		allocateFromFile(file, dataSize);
+	}
+
+	/**
+	 * Creates a copy of the span owned by the given owner.
+	 */
+	inline SpanOwner(const SpanOwner &other) {
+		// Allocating memory when copy-constructing from an unallocated owner
+		// will break the new owner by making it appear allocated even though
+		// it doesn't (and shouldn't) contain data
+		if (!other) {
+			SpanOwner();
+			return;
+		}
+		allocateFromSpan(other._span);
+	}
+
+	inline ~SpanOwner() {
+		free((void *)_span.data);
```

Comment:
This needs an explicit `const_cast`. (Or alternatively the allocate functions could store the pointer before having it cast to a `const T*`)